Mention copyleft in license guide! (#189)
pyOpenSci/python-package-guide: documentation/repository-files/license-files.md

[submitting your package to pyOpenSci for peer review](https://www.pyopensci.org/about-peer-review/index.html), then we require an OSI approved
license.

> :::{admonition} Copyleft licenses
> The other major category of licenses are ["copyleft" licenses](https://en.wikipedia.org/wiki/Copyleft).

Reviewer: ```suggestion
:::{admonition} Copyleft vs Permissive Licenses
While we suggest that you use a permissive license for your Python package, you may also see packages that use ["copyleft" licenses](https://en.wikipedia.org/wiki/Copyleft). 
```

Author: The previous line is "We generally suggest that you use a permissive license" so i think that it's reasonably clear copyleft is being presented as an alternative? i would worry that the language 'you may also see copyleft' is like 'you shouldn't use it but should be aware of it' rather than 'the other option you have is copyleft' - esp for new programmers i think it's fine to make a recommendation (that's the purpose of the guide!) but i wouldn't want it to come across as like 'don't use copyleft'. does that make sense? delicate balance i think

Reviewer: it makes sense @sneakers-the-rat let's leave as is for now. so sorry it's taken me 2 weeks to get back to this.

Reviewer 2: Marking this as resolved.

> These are licenses that require that anyone that uses your work redistributes it with the same (or greater) rights to modify, copy, share, and redistribute it.

Reviewer: ```suggestion
Copy left licenses requires people that use your work to redistribute it with the same (or greater) rights to modify, copy, share, and redistribute it.
```

Author: i like this, revise to "Copyleft licenses require" and i'll commit?

Reviewer: Fixed!

> In other words, copyleft licenses prohibit someone taking your work, making a proprietary version of it, and redistributing it without providing the source code so others can do the same.
> Copyleft licenses are "sticky" in that they are designed to ensure that more free software is created.

Reviewer: ```suggestion
Copyleft licenses are "sticky" in that they are designed to ensure that more free software is created. If your work has a copyleft license, it prohibits someone taking it, making a proprietary version of it, and redistributing it without providing the source code for others to use and share.

```

Reviewer: i'm not sure if this is quite right but just trying to simplify the words a bit :)

Author: Hmm ya you're right there needs to be a little more glue between these ideas bc it's not quite clear how they encourage more free software from this alone. I think the order of explaining what they are -> what 'stickiness' means makes sense, but maybe like "This requirement makes copyright licenses "sticky" - they are designed to encourage more free, copyleft licensed software to be created."

Reviewer: i think i follow!! i updated the revision.

> Copyleft licenses are "sticky," in that they are designed to ensure that more free software is created.

Reviewer 3: ```suggestion
Copyleft licenses are "sticky" in that they are designed to ensure that more free software is created.
```
@sneakers-the-rat Please disregard my suggestion if it's not a typo.

> The difference between copyleft vs. permissive licenses is an important cultural divide in free and open source software (eg. see {footcite}`hunterReclaimingComputingCommons2016`, {footcite}`gnuprojectWhatFreeSoftware2019`, {footcite}`gnuprojectWhatCopyleft2022`),

Reviewer 3: ```suggestion
The difference between copyleft vs. permissive licenses is an important cultural divide in free and open source software (e.g., see {footcite}`hunterReclaimingComputingCommons2016`, {footcite}`gnuprojectWhatFreeSoftware2019`, {footcite}`gnuprojectWhatCopyleft2022`),
```

> The difference between copyleft vs. permissive licenses is an important cultural divide in free and open source software (e.g., see {footcite}`hunterReclaimingComputingCommons2016`, {footcite}`gnuprojectWhatFreeSoftware2019`, {footcite}`gnuprojectWhatCopyleft2022`),
> that you should be aware of when choosing your license - the lineage of copyleft represents the "free" part of "free and open source software".

Reviewer: ```suggestion
The difference between copyleft and permissive licenses is an important cultural divide in free and open source software (e.g., see {footcite}`hunterReclaimingComputingCommons2016`, {footcite}`gnuprojectWhatFreeSoftware2019`, {footcite}`gnuprojectWhatCopyleft2022`).
It is important to understand this difference when choosing your license. Copyleft licenses represents the "free" part of "free and open source software".
```

> Free and open source software is intrinsically political, and it is important to be aware of power dynamics in computing as well as the practical problems of license compatibility (discussed below).

Reviewer: This is al great @sneakers-the-rat , thank you!! I have a few thoughts and questions here. 

1. i think for someone who is new to software and OSS the political part will seem "out of no where" and potentially scary even tho it is truth. i wonder if there is a way to make this statement a bit more clear to someone that is new. i feel like there is a leap here from - let's add a license to your package to - there are power dynamics at play and licenses can be play a role in these dynamics. i'm not sure what language we should use but i just wonder if we can make this more clear to someone who is newer to the space. It's especially confusing as projects like numpy ahd choosealicense.org suggest that we use MIT. so i'd be asking - well it seems like everyone should use GNU then. Why are we using MIT? why are we suggesting MIT.
do you see where i'm going here? can we adjust to consider those types of questions and thoughts / concerns? (honestly i'm thinking the same right now as we decided on MIT as a community but copyleft does sound compelling the way it's described here.

Author: Maybe we can say something like "This set of tutorials is primarily intended to teach you to program, but an important part of learning to be a programming is learning about the history and culture of open source software. Licensing is a decision about who can do what with your code, which is intrinsically political ... {rest of text} ... there are legitimate differences in goals and context for open source software, and each category of licenses has their uses, supporters, and detractors"

Reviewer: i think that makes a lot of sense to frame it . do you want to make that change to the file? This is also making me think that we should either have a blog post on this topic (probably preferred) OR add a page to the guide about the culture of open source software. Or both. that way we can link to that content rather than adding too much to individual pages. how about that?  i can open this as a new issue in this repo and we can decide whether it should be a blog or a  page in the guide. and maybe we can get the community to help write it if someone starts it. 

i'll open the issue now and we can discuss separately

process.

:::{admonition} License recommendations from the SciPy package
[The SciPy documentation has an excellent overview of licenses.](https://docs.scipy.org/doc/scipy/dev/core-dev/index.html#licensing). One of the key elements
that these docs recommend is ensuring that the license that you select is
compatible with licenses used in many parts of the scientific Python ecosystem.
Below is a highlight of this text which outlines license that are compatible
with the modified BSD license that SciPy uses.

> Other licenses that are compatible with the modified BSD license that SciPy uses are 2-clause BSD, MIT and PSF. Incompatible licenses are GPL, Apache and custom licenses that require attribution/citation or prohibit use for commercial purposes.

> If your primary goal is for your code to be used by other, major packages in the scientific ecosystem, we also recommend

Reviewer: see this conflicts with the breakout above. and i totally understand why we want that breakout above.  but a new user will be confused by this after reading about copyleft! so how can we create a text bridge between the two?

Author: how do you mean? like we should have a parallel "otherwise, if your goal is..." for copyleft? I read it as previous section recommending permissive licenses, and then this section further explaining why that would be the case - there are lots of major packages that use permissive, and if you use copyleft then your work can't be used in them. can you tell me more about what's reading as discordant?

sure that the license for the code that you are using also complies
with the license that you selected for your package.

> A useful way to think about license compatibility is the distinction between **"inbound"** and **"outbound"** compatibility.

Reviewer: i feel like some of this would be useful above. what is we thought about re-organizing this page jonny? 

at the top we have 

## Where to store your license

### Use open permissive licenses when possible

what is instead we do something like this

```
## Where to store your license
keep text as is 

##Types of licenses: permissive vs copyleft 

explain inbound vs outbound restrictions here to frame the text below

### Overview of permissive licenses

### Overview of copyleft licenses (make your breakout into this section)

 ## How to chose a license 

* we suggest mit or bsd-3 for compatibilty with the scientific Python ecosystem
* scipy text as a reference 

## Following license guidelines

more about how to follow and how that impacts permissive vs copyleft (which by now it's clear how these are different). and yo've already explain inbound vs outbound permisions above so you can just have a little table even and demonstrate how the licenses are different.

Author: I think this should probably come after we give definitions of what the categories of licenses are, but i'm definitely game to making a separate section for each type, that is def clearer and might help above comment re: ordering. 

I see this thought as being part of all three 'what are the types of licenses,' 'why might i choose one vs another,' and 'how do i follow license guidelines' so idk! I think it's nicely paired with the example of stack overflow. so i could see this part going one of a few places...

```
- Where to store license
- Types of licenses
  - permissive
  - copyleft
  - comparison <--- maybe here???
- choosing a license
  - suggestion
  - scipy thing
  - considerations to weigh <--- maybe here???
- following license guidelines
  - when you can and can't use software given your license <--- maybe here???
  - stack overflow example
```

i don't rly have a strong preference, but i think you're right maybe sooner is good :)

Reviewer: i think what you have above is fine. i just think all of the overview should be in one section and then choosing should be it's own section with limited additional info but definitely a considerations section

Reviewer 2: Opened issue #415.

"Inbound" licenses are those that cover the software you plan to include in your package.
Your package is protected by an "outbound" license.

**Permissive licenses** like BSD and MIT have few **outbound** restrictions - they can be used in any way by downstream consumers, including making them proprietary.
This is why they are favored by many businesses and large packages that want to be adopted by businesses.
Permissive licenses have more **inbound** restrictions - they can't use software that requires more freedoms to be preserved than they do, like copyleft licenses.
A package licensed under MIT needs to take special care when including or modifying a package licensed under the GPL-3.

**Copyleft licenses** like GPL-3 have more **outbound** restrictions - they require more of packages that include, use, modify, and reproduce them.
This is the purpose of copyleft licenses, to ensure that derivative works remain free and open source.
They have fewer **inbound** restrictions - a GPL-3 licensed package can include any other permissively licensed and most copyleft licensed packages.

> |                                                      Compatible | Dependency <br> ("Inbound") | Your Package | Downstream Package <br> ("Outbound") |

Reviewer: OMG you have a table - yes! i was thinking a table would be great. [myst tables](https://myst-parser.readthedocs.io/en/latest/syntax/tables.html) are pretty create btw. simpler syntax / easier to read :)

Author: you mean the list table directive? i can swap it to that if you'd prefer :)

Reviewer 2: Opened #416 as a task for post-merge.

| <i class="fa-solid fa-check" style="color: MediumSeaGreen"></i> | Copyleft                    | Copyleft     |                                      |
|            <i class="fa-solid fa-x" style="color: Crimson"></i> |                             | Copyleft     | Permissive                           |
| <i class="fa-solid fa-check" style="color: MediumSeaGreen"></i> |                             | Copyleft     | Copyleft                             |

:::{admonition} An example of how a license determine how code can be reused
:class: note

Let's use StackOverflow as an example that highlights how a license determines how code can or can not be used.

[Stack Overflow uses a Creative Commons Share Alike license.](https://stackoverflow.com/help/licensing). The sharealike license requires you to use the same sharealike license when you reuse any code from Stack Overflow.

This means that from a legal perspective, if you copy code from the Stack Overflow website and use it in your package that is licensed differently, say with a MIT license, you are violating Stack Overflow's license requirements!

> This would not be true with a GPL licensed package - `GPL-3` packages can include code licensed by `CC-BY-SA` {footcite}`creativecommonsShareAlikeCompatibilityGPLv32015`.

Reviewer: ```suggestion
This would not be true with a GPL licensed package. `GPL-3` packages can include code licensed by `CC-BY-SA` {footcite}`creativecommonsShareAlikeCompatibilityGPLv32015`.
```

[Stack overflow uses a Creative Commons Share Alike license.](https://stackoverflow.com/help/licensing). The sharealike license requires you to use the same sharealike license when you reuse any code from stackoverflow.

> This means that technically, if you copy code from the Stack Overflow website, and use it in your package. And your packages uses a different license such as a MIT license, you are violating Stack Overflow's license requirements!

Author: thanks for these suggestions, merging now<3
